### ISSUE TITLE
Objects vanish after a while

### ISSUE THREAD

**Issue by adammal96** [samp-incognito/samp-streamer-plugin#230]
We are running into an issue where after a couple hours of our server running any newly placed or newly edited objects vanish directly after spawn. We have tried changing the settings back to default yet nothing has helped, we have roughly 250 static objects used as bases for when vehicles spawn on them, and this happens all over the map not just near other mapping where the object limit may be reached. We are kind of at a loss here as nothing really seems to be working.

**Comment by IstuntmanI** [comment]
I had a similar bug when I used v2.9.0 for a short time (then gone back to v2.8.2). Pickups disappeared so I had to restart the server fast. You are using v2.9.0+, right ?

Probably related to the chunk/per-player tick changes too. Those were really major changes which weren't tested too much.

**Comment by adammal96** [comment]
yes I am using the latest version of the plugin, I'll take a look at going back to 2.8.2 to see if that resolves anything

**Comment by IstuntmanI** [comment]
Can you try the latest commit ? I added the files [here](https://github.com/samp-incognito/samp-streamer-plugin/issues/205#issuecomment-329974272).

**Comment by adammal96** [comment]
Will do.

**Comment by raydex23** [comment]
Something is really wrong. I've got similar problems since update to the latest stable version.
Newly placed objects are dissapearing after while, and their IDs are messed up. 
GetItemStreamerID is returning invalid object id even if i still see object.

**Comment by IstuntmanI** [comment]
You should try the files I added for downloading too.

**Comment by raydex23** [comment]
3 days of testing your file, nothing changed. Same problems.

**Comment by adammal96** [comment]
We've been testing it as well, our objects still vanish after ~ 24 hours

**Comment by adammal96** [comment]
This also happens on version 2.8.2 as well, albeit it takes slightly longer. It only happens to objects that aren't spawned on server load, this also happens to objects that are edited. If I spawn a new object like a roadblock, they will vanish after a couple minutes, if I load a gate object on server start, it won't vanish unless I edit it, again with a couple minutes of staying. This also applies to Dynamic 3d text, if its moved it vanishes in a couple minutes

**Comment by IstuntmanI** [comment]
That's weird. I'm using 2.8.2 on my server and it has no problem. Are you sure that you loaded 2.8.2 and not 2.9+ ? AFAIK nobody has similar problems with 2.8.2.

**Comment by raydex23** [comment]
2.8.2 is fine, must be your scripting bug.

**Comment by adammal96** [comment]
Anything I should be looking at? I’m not sure what exactly could be wrong,

**Comment by samp-incognito** [comment]
2.8.2 is fine. This is the first I've heard of this happening in 2.9.1, though. Make sure you aren't mixing up the IDs in your own script (e.g., when you destroy an object, make sure you reset the variable pointing to that object in your script).

**Comment by adammal96** [comment]
I've made sure I'm not mixing up ID's, it doesn't occur until after ~24+ hours of the server constantly running which puzzles me because if it was mixing up ID's it would happen all the time.

**Comment by IstuntmanI** [comment]
Can you give us more details of the moment when it happens ?:
- return value of `Streamer_CountVisibleItems` (`serverwide = 1`; the container may fail to get freed for some reason and the internal visible items count be reached and stop further streaming ?)
- press F5 in-game and tell us the values of "RWObjects" and `ObjectSlotsUsed` (well, these should be normal, but just in case, I guess)
- can you try a native `CreatePlayerObject` and see if it works ? (well, if the previous point's values are regular, this should work too)

These are the main reasons the streaming loop may stop, so they may help.

**Comment by adammal96** [comment]
I'll get back to you on the other items  as those will require me to push an update however, 
RWObjects is 877
ObjectSlotsUsed is 374
In pershing square by lots of mapping its at
RWObjects 952
ObjectSlotsUsed 647

**Comment by adammal96** [comment]
The streamer CountVisibleItems is around 100-200 when dynamic objects are disappearing, creating a static object results in it staying. It doesn't disappear like dynamics do.

**Comment by IstuntmanI** [comment]
That's weird. Maybe Incognito can see something suspicious in that behavior, probably at the deleting part.

By the way, it looks like it isn't happening for all items at once, but it happens to random items types. It may happen only to objects, or only to pickups, or only to objects, most probably for more than one item type too.

**Comment by adammal96** [comment]
Yea we can’t pin it on any specific period of time that this occurs as it
doesn’t become noticeable until around 24 hours. it generally doesn’t
happen to pickups for us as well only objects as far as I’ve noticed

**Comment by spacemud** [comment]
I'm experiencing a remarkably similar problem, also on 2.8.2. The objects don't spontaneously disappear for me, but rather get destroyed when other objects are spawned.

I've put debug prints at every point of my code where objects are created and destroyed and there is nothing to suggest that they are being destroyed from the script. The newly spawned objects often have the same ID as an existing object, suggesting that they are being 'overwritten' somehow. I'm very diligent about resetting data when I destroy an object, so doubt it is that. I will recheck though.

I will try some of the suggestions offered in this topic and report back.

**Comment by adammal96** [comment]
Just gonna bump this, still having this issue, I'm willing to try just about anything here. I may try what spacemud did and see if its overwriting.

**Comment by IstuntmanI** [comment]
The best idea would be to add tons of prints in functions like `processObjects`, `processMapIcons` and so on I guess. I think that some temporary `Streamer_EnableItemCreateDebug` would be nice (maybe also with a type parameter ? AFAIK not all item types start being bugged, but only one of them could get bugged, or more), so when you are no longer seeing any item type, you could enable it with some command and see tons of prints in the console and inspect it. Those prints should be placed at the start of `processObjects`, to see if it is actually checked, also after the first loop, before the creation (after discovering objects) and at the end to see if it actually ended (well, it should be printed always if the first one is printed) and of course some prints in those loops, also before some `break;` lines, also printing in those loops the sizes of `player.internalObjects`, `discoveredObjects`, `existingObjects` and that's all probably. If the first one isn't printed, then we should advance the debugging procedure to the `Streamer::performPlayerUpdate` function, but I guess that `processObjects` is actually called. Also, as I said in my first reply here:
`I had a similar bug when I used v2.9.0 for a short time (then gone back to v2.8.2). Pickups disappeared so I had to restart the server fast. You are using v2.9.0+, right ?`
the functions for pickups and actors should probably be debugged too, in case that the bug happens on those item types instead, even if they are globally streamed and not just per-player.

So yeah, the biggest problem is that no one knows why this happens and we can't easily add prints all over it, because we have to wait a lot of time. A function to enable those debugs would be nice so the console won't be spammed for every player when it is not needed.

That's what we need: someone to use it and debug everything he can.

**Comment by adammal96** [comment]
I've started the process of adding print statements with object ID's, I will also do this for 3d text and pickups as well. I'll test this on 2.8.2 first and move to 2.9.0 if I find a result.

**Comment by IstuntmanI** [comment]
v2.8.2 is fine, all you need to do is check them in the latest version and see what seems suspect when items stop appearing.

**Comment by IstuntmanI** [comment]
@samp-incognito I think I found the problem... if this behavior isn't actually the correct behavior. I'll submit here all warnings I got:
```
==== Building streamer (debug) ====
streamer.cpp
In file included from src/core.h:22:0,
                 from src/streamer.cpp:19:
src/grid.h: In member function ‘void Grid::eraseCellIfEmpty(const SharedCell&)’:
src/grid.h:94:2: warning: declaration of ‘cell’ shadows a global declaration [-Wshadow]
  {
  ^
In file included from lib/sampgdk/sampgdk.h:165:0,
                 from src/common.h:76,
                 from src/cell.h:20,
                 from src/streamer.h:20,
                 from src/streamer.cpp:17:
lib/samp-plugin-sdk/amx/amx.h:151:21: note: shadowed declaration is here
   typedef int32_t   cell;
                     ^~~~
src/streamer.cpp: In member function ‘void Streamer::processMapIcons(Player&, const std::vector<boost::intrusive_ptr<Cell> >&)’:
src/streamer.cpp:876:47: warning: declaration of ‘i’ shadows a previous local [-Wshadow]
      boost::unordered_map<int, int>::iterator i = player.internalMapIcons.find(e->second->mapIconID);
                                               ^
src/streamer.cpp:864:44: note: shadowed declaration is here
   boost::unordered_map<int, int>::iterator i = player.internalMapIcons.find(d->second->mapIconID);
                                            ^
src/streamer.cpp: In member function ‘void Streamer::processObjects(Player&, const std::vector<boost::intrusive_ptr<Cell> >&)’:
src/streamer.cpp:981:46: warning: declaration of ‘i’ shadows a previous local [-Wshadow]
     boost::unordered_map<int, int>::iterator i = player.internalObjects.find(d->second->attach->object);
                                              ^
src/streamer.cpp:971:44: note: shadowed declaration is here
   boost::unordered_map<int, int>::iterator i = player.internalObjects.find(d->second->objectID);
                                            ^
src/streamer.cpp:996:47: warning: declaration of ‘i’ shadows a previous local [-Wshadow]
      boost::unordered_map<int, int>::iterator i = player.internalObjects.find(e->second->objectID);
                                               ^
src/streamer.cpp:971:44: note: shadowed declaration is here
   boost::unordered_map<int, int>::iterator i = player.internalObjects.find(d->second->objectID);
                                            ^
src/streamer.cpp: In member function ‘void Streamer::processTextLabels(Player&, const std::vector<boost::intrusive_ptr<Cell> >&)’:
src/streamer.cpp:1291:47: warning: declaration of ‘i’ shadows a previous local [-Wshadow]
      boost::unordered_map<int, int>::iterator i = player.internalTextLabels.find(e->second->textLabelID);
                                               ^
src/streamer.cpp:1279:44: note: shadowed declaration is here
   boost::unordered_map<int, int>::iterator i = player.internalTextLabels.find(d->second->textLabelID);
                                            ^
amx.cpp
In file included from src/utility/../core.h:22:0,
                 from src/utility/amx.cpp:19:
src/utility/../grid.h: In member function ‘void Grid::eraseCellIfEmpty(const SharedCell&)’:
src/utility/../grid.h:94:2: warning: declaration of ‘cell’ shadows a global declaration [-Wshadow]
  {
  ^
In file included from lib/sampgdk/sampgdk.h:165:0,
                 from src/utility/../common.h:76,
                 from src/utility/misc.h:20,
                 from src/utility/amx.h:20,
                 from src/utility/amx.cpp:17:
lib/samp-plugin-sdk/amx/amx.h:151:21: note: shadowed declaration is here
   typedef int32_t   cell;
                     ^~~~
src/utility/amx.cpp: In function ‘void Utility::executeFinalAreaCallbacks(int)’:
src/utility/amx.cpp:219:33: warning: declaration of ‘a’ shadows a previous local [-Wshadow]
   for (std::set<AMX*>::iterator a = core->getData()->interfaces.begin(); a != core->getData()->interfaces.end(); ++a)
                                 ^
src/utility/amx.cpp:205:56: note: shadowed declaration is here
  boost::unordered_map<int, Item::SharedArea>::iterator a = core->getData()->areas.find(areaid);
                                                        ^
geometry.cpp
In file included from src/utility/../core.h:22:0,
                 from src/utility/geometry.cpp:20:
src/utility/../grid.h: In member function ‘void Grid::eraseCellIfEmpty(const SharedCell&)’:
src/utility/../grid.h:94:2: warning: declaration of ‘cell’ shadows a global declaration [-Wshadow]
  {
  ^
In file included from lib/sampgdk/sampgdk.h:165:0,
                 from src/utility/../common.h:76,
                 from src/utility/misc.h:20,
                 from src/utility/geometry.h:20,
                 from src/utility/geometry.cpp:17:
lib/samp-plugin-sdk/amx/amx.h:151:21: note: shadowed declaration is here
   typedef int32_t   cell;
                     ^~~~
misc.cpp
In file included from src/utility/../core.h:22:0,
                 from src/utility/misc.cpp:19:
src/utility/../grid.h: In member function ‘void Grid::eraseCellIfEmpty(const SharedCell&)’:
src/utility/../grid.h:94:2: warning: declaration of ‘cell’ shadows a global declaration [-Wshadow]
  {
  ^
In file included from lib/sampgdk/sampgdk.h:165:0,
                 from src/utility/../common.h:76,
                 from src/utility/misc.h:20,
                 from src/utility/misc.cpp:17:
lib/samp-plugin-sdk/amx/amx.h:151:21: note: shadowed declaration is here
   typedef int32_t   cell;
                     ^~~~
```

My suggestions:
1. (unrelated) Change `src/grid.h` `eraseCellIfEmpty`'s function argument name from `cell` to something like `sharedCell`, `localCell`, `passedCell` or whatever you like it to be called like, `cell` is already defined by the SDK to be `int`, so there's shadowing.
2. *The most important ones, related to this problem*: Fix all the other shadow warnings (mainly from `src/streamer.cpp`). I told you some time ago do so, but now I inspected them a bit and they are really looking like a problem. Let's look at one of them:
```
src/streamer.cpp: In member function ‘void Streamer::processObjects(Player&, const std::vector<boost::intrusive_ptr<Cell> >&)’:
src/streamer.cpp:981:46: warning: declaration of ‘i’ shadows a previous local [-Wshadow]
     boost::unordered_map<int, int>::iterator i = player.internalObjects.find(d->second->attach->object);
                                              ^
src/streamer.cpp:971:44: note: shadowed declaration is here
   boost::unordered_map<int, int>::iterator i = player.internalObjects.find(d->second->objectID);
                                            ^
src/streamer.cpp:996:47: warning: declaration of ‘i’ shadows a previous local [-Wshadow]
      boost::unordered_map<int, int>::iterator i = player.internalObjects.find(e->second->objectID);
                                               ^
src/streamer.cpp:971:44: note: shadowed declaration is here
   boost::unordered_map<int, int>::iterator i = player.internalObjects.find(d->second->objectID);
                                            ^
```
Here's the code:
https://github.com/samp-incognito/samp-streamer-plugin/blob/4efb5055a6caf82eaa33434295e2a4c917cd07e1/src/streamer.cpp#L971-L1013

You can see that the iterator `i` is created 3 times in the same scope (well, the first one has a bigger scope than the rest) with the same type: `boost::unordered_map<int, int>::iterator`, so everytime `i` is used it will get its latest value. I guess that some of those IDs get messed up and it isn't removing/searching what it should. I'm not exactly sure what happens there because it is pretty messy and I couldn't track the code too well because the code was changed a lot over time (I tried looking in an older version's code, but it is too different to know how it should look like). I think that you will know how these should be fixed.
Also, the other shadow warnings (for mapIcons and textLabels) are mostly the same, excepting that they have only one warning each instead of two. This may be why users got problems with the objects more ?

After you will do this, I may create another pull request to add a bigger amount of warnings when compiling with GCC and ignoring the ones from Boost, Eigen, sampGDK and SDK. If you want, you can do this in the next commit, all you have to do is modify a bit the `premake5.lua` file and regenerate the make files for `vs2015` and `gmake`:
1. Change 
`includedirs { "include", "lib", "lib/**" }`
to
`sysincludedirs { "include", "lib", "lib/**" }`
( https://stackoverflow.com/questions/1867065/how-to-suppress-gcc-warnings-from-library-headers )
2. Add
`buildoptions { "-Wextra", "-pedantic", "-Wconversion", "-Wcast-align", "-Wcast-qual", "-Wctor-dtor-privacy", "-Wdisabled-optimization", "-Wformat=2", "-Winit-self", "-Wlogical-op", "-Wmissing-include-dirs", "-Wnoexcept", "-Woverloaded-virtual", "-Wredundant-decls", "-Wshadow", "-Wsign-promo", "-Wstrict-null-sentinel", "-Wstrict-overflow=5", "-Wundef", "-Wno-unused", "-Wno-variadic-macros", "-Wno-parentheses", "-fdiagnostics-show-option" }`
after
`targetextension ".so"`
and it should work just fine. Showing all the right errors.

I didn't send a pull request for this because I don't really know which `i` should be replaced with other name and which not. Also because you may want another name for that function argument.

**Comment by samp-incognito** [comment]
I really doubt the shadowed variables are causing any problems in those specific cases. For clarity, though, yes, their names should be changed to something else.

**Comment by adammal96** [comment]
So the issue is currently happening. Still can't figure out why.

Line 85399: [09/12/2017 19:49:20] Generating Dyn Object 16477
Line 85405: [09/12/2017 19:49:36] Generating Dyn Object 16478

It appears to be assigning valid object ID's but they just disappear, they aren't getting overwritten or anything.

**Comment by IstuntmanI** [comment]
@samp-incognito Well, I just thought that you are using the wrong variable because of the shadowing. I know that shadowing is legal. Well, I may do a pull request with everything I said.

@adammal96 If by "generating" you mean that `CreatePlayerObject` is actually called fine (so this means that they are discovered fine), then you should check if `DestroyPlayerObject` is called immediately after it. How often is this happening now ?

**Comment by adammal96** [comment]
Happens every day, but only after like 15-20 hours. It prints the object ID when doing CreateDynamicObject, and prints when destroying but it doesn't destroy it, the object just vanishes

**Comment by IstuntmanI** [comment]
I don't really get what you mean with the destroying part. If the `DestroyPlayerObject` part of code is called, it should be destroyed like it needs to. You mean that part isn't called at all ? Well, the object can't just vanish without any call to `DestroyPlayerObject`.

**Comment by adammal96** [comment]
As far as I can see it doesn't get called. It just vanishes.
